lib/Runtime/Library/InJavascript/Intl.js

```diff
@@ -552,54 +561,86 @@
 
-    /**
-     * Returns an object representing the language, script, region, extension, and base of a language tag,
-     * or null if the language tag isn't valid.
-     *
-     * @param {String} langtag a candidate BCP47 langtag
-     */
-    const parseLangtag = function (langtag) {
-        const parts = _.match(langtag, LANG_TAG_RE);
-        if (!parts || !parts[1]) {
-            return null;
-        }
+        const parsedLangtagCache = new IntlCache();
+        return function (langtag) {
+            const cached = parsedLangtagCache.get(langtag);
+            if (cached) {
+                return cached;
+            }
 
-        const ret = { language: parts[1], base: parts[1] };
-        if (parts[2]) {
-            ret.script = parts[2];
-            ret.base += "-" + parts[2];
-        }
+            const parts = _.match(langtag, LANG_TAG_RE);
+            if (!parts) {
+                return null;
+            }
 
-        if (parts[3]) {
-            ret.region = parts[3];
-            ret.base += "-" + parts[3];
-        }
+            const ret = _.create();
+            ret.language = parts[1];
+            ret.base = parts[1];
+            if (parts[2]) {
+                ret.script = parts[2];
+                ret.base += "-" + parts[2];
+            }
 
-        if (parts[4]) {
-            // strip the first character to turn -u-... into u-...
-            ret.extension = _.substring(parts[4], 1);
-        }
+            if (parts[3]) {
+                ret.region = parts[3];
+                ret.base += "-" + parts[3];
+            }
 
-        return ret;
-    }
+            if (parts[4]) {
+                ret.variants = parts[4];
+            }
 
-    function IsWellFormedCurrencyCode(code) {
+            if (parts[5]) {
+                ret.extensions = parts[5];
+
+                // parse the extension to find the unicode (-u) extension
+                const extensionParts = _.split(parts[5], "-");
+                for (let i = 0; i < extensionParts.length; ++i) {
+                    if (extensionParts[i] !== "u") {
+                        continue;
+                    }
+
+                    let k;
+                    for (k = i + 1; k < extensionParts.length && extensionParts[k].length > 1; k++) {
+                        // do nothing, we just want k to equal the index of the next element whose length is 1
+                        // or to equal the length of extensionParts
+                        // We could have done this with Array.prototype.findIndex too
+                    }
+
+                    if (k > i + 1) {
+                        // this creates u-(keys and values)*, which is good enough for the UnicodeExtensionValue,
+                        // which is the only place that this return value is intended to be used
+                        ret.unicodeExtension = _.join(_.slice(extensionParts, i, k), "-");
```

Comment:
>unicodeExtension [](start = 28, length = 16)

why not store this as a dictionary of keys to values and then reconstitute the string later when needed?

Reply:
We could, but then we would be obviating/reimplementing UnicodeExtensionValue here, which would bring us further away from spec.